File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        def rec(i, ds, summ):
            if i==len(candidates):
                return
            if summ ==  target:
                ans.append(ds[:])
                return
            if summ+candidates[i]<=target:    
                ds.append(candidates[i])
                summ+=candidates[i]
                rec(i, ds, summ)
                summ-=candidates[i]
                ds.pop()
            rec(i+1, ds, summ)  
        rec(0, [], 0)
        return ans          

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        backtrack(candidates, target, 0, new ArrayList<>(), 0, ans);
        return ans;
    }
    private void backtrack(int[] candidates, int target, int i, List<Integer> ds, int summ, List<List<Integer>> ans){
        if (i==candidates.length){
            return;
        }
        if (summ==target){
            ans.add(new ArrayList<>(ds));
            return;
        }
        if (summ+candidates[i]<=target){
            ds.add(candidates[i]);
            summ+=candidates[i];
            backtrack(candidates, target, i, ds, summ, ans);
            summ-=candidates[i];
            ds.remove(ds.size()-1);
        }
        backtrack(candidates, target, i+1, ds, summ, ans);
    }
}